test/e2e/util.go

```diff
@@ -0,0 +1,599 @@
+	case "POST":
+		reqHeader.Set("Content-Type", "application/x-www-form-urlencoded")
+		reqBody = strings.NewReader(formData.Encode())
+	default:
+		return nil, fmt.Errorf("unknown method: %s", reqMethod)
+	}
+
+	req, err := http.NewRequest(reqMethod, reqURL.String(), reqBody)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header = reqHeader
+	return req, nil
+}
+
+type RouteTemplate struct {
+	AppLabel   string
+	NameLabel  string
+	TargetPort int32
+	ToName     string
+	Namespace  string
+}
+
+var routeYaml = `apiVersion: v1
+kind: Route
+metadata:
+  labels:
+    app: {{.AppLabel}}
+  name: {{.NameLabel}}
+  namespace: {{.Namespace}}
+spec:
+  port:
+    targetPort: {{.TargetPort}}
+  to:
+    kind: Service
+    name: {{.ToName}}
+    weight: 100
+  wildcardPolicy: None
+  tls:
+    termination: passthrough
+`
+
+// execCmd executes a command and returns the stdout + error, if any
+func execCmd(cmd string) (string, error) {
+	parts := strings.Fields(cmd)
+	head := parts[0]
+	parts = parts[1:]
+
+	out, err := exec.Command(head, parts...).CombinedOutput()
+	if err != nil {
+		fmt.Printf("Command '%s' failed with: %s\n", cmd, err)
+		fmt.Printf("Output: %s\n", out)
+		return "", err
+	}
+	return string(out), nil
+}
+
+func DeleteTestRoute(routeName string) error {
+	_, err := execCmd(fmt.Sprintf("oc delete route/%s", routeName))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func GetRouteHost(routeName, namespace string) (string, error) {
+	out, err := execCmd(fmt.Sprintf("oc get route/%s -o jsonpath='{.spec.host}' -n %s", routeName, namespace))
+	if err != nil {
+		return "", err
+	}
+	// strip surrounding single quotes
+	return out[1 : len(out)-1], nil
+}
+
+func NewOAuthProxyService(serviceName, appLabel string, port int32, targetPort int) *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: serviceName,
+			Labels: map[string]string{
+				"app": appLabel,
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app": appLabel,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Protocol:   corev1.ProtocolTCP,
+					Port:       port,
+					TargetPort: intstr.FromInt(targetPort),
+				},
+			},
+		},
+	}
+}
+
+// create a route using oc create directly
+func NewOAuthProxyRoute(app, name, toname, namespace string, port int32) error {
+	tmpl, err := template.New("route").Parse(routeYaml)
+	if err != nil {
+		return err
+	}
+
+	routeYamlFile, err := ioutil.TempFile(os.TempDir(), "proxy-test-route")
+	if err != nil {
+		return err
+	}
+	defer routeYamlFile.Close()
+	defer os.Remove(routeYamlFile.Name())
+
+	// execute template
+	writer := bufio.NewWriter(routeYamlFile)
+	err = tmpl.Execute(writer, RouteTemplate{
+		AppLabel:   app,
+		NameLabel:  name,
+		TargetPort: port,
+		ToName:     toname,
+		Namespace:  namespace,
+	})
+
+	if err != nil {
+		return err
+	}
+	writer.Flush()
+
+	// create route
+	_, err = execCmd(fmt.Sprintf("oc create -f %s", routeYamlFile.Name()))
```

Comment:
If you're going to exec commands, pass arguments as `[]string`, not as a string.  And use stdin instead of writing to a temporary file.

Is the templating really worth it here vs just creating it in code?

Reply:
The setup details like app name, route name, port, etc. I thought might be relevant to some test scenarios and we would need to be able to change them per-test. for now, though, they could be constant.